File: ui/openmole-ui-console/org.openmole.ui.console/src/main/java/org/openmole/ui/console/Application.java
package org.openmole.ui.console;

import java.io.File;
import java.util.logging.Logger;
import org.apache.commons.cli.BasicParser;
import org.apache.commons.cli.CommandLine;
import org.apache.commons.cli.CommandLineParser;
import org.apache.commons.cli.HelpFormatter;
import org.apache.commons.cli.Option;
import org.apache.commons.cli.OptionBuilder;
import org.apache.commons.cli.Options;
import org.apache.commons.cli.ParseException;
import org.codehaus.groovy.tools.shell.Groovysh;
import org.eclipse.equinox.app.IApplication;
import org.eclipse.equinox.app.IApplicationContext;
import org.openmole.misc.pluginmanager.PluginManager;
import org.openmole.misc.workspace.Workspace;
import org.openmole.ui.console.internal.command.Get;
import org.openmole.ui.console.internal.command.Init;
import org.openmole.ui.console.internal.command.Print;
import org.openmole.ui.console.internal.command.Auth;
import org.openmole.ui.console.internal.command.Encrypted;

public class Application implements IApplication {


    @Override
    public Object start(IApplicationContext context) throws Exception {

        // Init options parsing
        String[] args = (String[]) context.getArguments().get("application.args");
        Option optionPluginsDir = OptionBuilder.withLongOpt("pluginsDir").withDescription("Add plugins directories (seperated by \",\")").hasArgs(1).withArgName("directories").isRequired(false).create("p");
        Option workspaceDir = OptionBuilder.withLongOpt("workspaceDir").withDescription("Directory of the workspace").hasArgs(1).withArgName("directory").isRequired(false).create("w");

        Options options = new Options().addOption(optionPluginsDir).addOption(workspaceDir);
        CommandLineParser parser = new BasicParser();
        
        CommandLine cmd = null;
        try {
            cmd = parser.parse(options, args);
        } catch (ParseException e) {
            Logger.getLogger(Application.class.getName()).severe("Error while parsing command line arguments");
            new HelpFormatter().printHelp(" ", options);
            return IApplication.EXIT_OK;
        }

        
        File workspaceLocation = Workspace.defaultLocation();
        if (cmd.hasOption(workspaceDir.getOpt())) {
            workspaceLocation = new File(cmd.getOptionValue(workspaceDir.getOpt()));
        }

        if(Workspace.anotherIsRunningAt(workspaceLocation)) {
            Logger.getLogger(Application.class.getName()).severe("Application is already runnig at " + workspaceLocation.getAbsolutePath()+ ". If it is not the case please remove the file '" + new File(workspaceLocation, Workspace.running()).getAbsolutePath() + "'.");
            return IApplication.EXIT_OK;
        }

        if(cmd.hasOption(workspaceDir.getOpt())) Workspace.instance_$eq(new Workspace(workspaceLocation));

        
        Groovysh g = Console.groovysh();
        Groovysh muteShell = Console.muteGroovysh();
        g.leftShift(new Print(g, "print", "\\pr"));
        g.leftShift(new Init(g, muteShell, "init", "\\in"));
        g.leftShift(new Get(g, muteShell, "get", "\\g"));
        g.leftShift(new Auth(g, muteShell, "auth", "\\au"));
        g.leftShift(new Encrypted(g, muteShell, "encrypted", "\\en"));
         
        Console.run("init " + Console.workspace());
        
        // Process CLI options
        if (cmd.hasOption(optionPluginsDir.getOpt())) {
            for (String directory : cmd.getOptionValue(optionPluginsDir.getOpt()).split(",")) {
                PluginManager.loadDir(directory);
            }
        }
        
        // Run
        Console.groovysh().run();
        return IApplication.EXIT_OK;
    }

    @Override
    public void stop() {
    }
}

File: ui/openmole-ui-console/org.openmole.ui.console/src/main/scala/org/openmole/ui/console/Application.scala
/*
 * Copyright (C) 2012 reuillon
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package org.openmole.ui.console

import java.io.File
import org.eclipse.equinox.app.IApplication
import org.eclipse.equinox.app.IApplicationContext
import org.openmole.misc.pluginmanager.PluginManager
import org.openmole.misc.tools.service.Logger
import org.openmole.misc.workspace.Workspace
import org.openmole.ui.console.internal.command.Auth
import org.openmole.ui.console.internal.command.Encrypt
import org.openmole.ui.console.internal.command.Get
import org.openmole.ui.console.internal.command.Print
import org.openmole.ui.console.internal.command.Verify
import scopt.immutable._

class Application extends IApplication with Logger {
  override def start(context: IApplicationContext) = {
    
    case class Config(
      pluginsDirs: List[String] = Nil,
      workspaceDir: Option[String] = None
    )
    
    
    val parser = new OptionParser[Config]("openmole-console", "0.x") { 
      def options = Seq(
        opt("p", "pluginDirectories" ,"Plugins directories (seperated by \",\")") {
          (v: String, c: Config) => c.copy(pluginsDirs = v.split(',').toList) 
        },
        opt("w", "Path of the workspace") {
          (v: String, c: Config) => c.copy(workspaceDir = Some(v)) 
        }
      )
    }
      
    parser.parse(context.getArguments.get("application.args").asInstanceOf[Array[String]], Config()) foreach { config =>
    
      val workspaceLocation = config.workspaceDir match {
        case Some(w) => new File(w)
        case None => Workspace.defaultLocation
      }

      if(Workspace.anotherIsRunningAt(workspaceLocation)) 
        logger.severe("Application is already runnig at " + workspaceLocation.getAbsolutePath + ". If it is not the case please remove the file '" + new File(workspaceLocation, Workspace.running).getAbsolutePath() + "'.")  
      else {       
        if(config.workspaceDir.isDefined) Workspace.instance = new Workspace(workspaceLocation)
 
        val g = Console.groovysh
        g.leftShift(new Print(g, "print", "\\pr"))
        g.leftShift(new Get(g, "get", "\\g"))
        g.leftShift(new Auth(g, "auth", "\\au"))
        g.leftShift(new Encrypt(g, "encrypt", "\\en"))
        g.leftShift(new Verify(g, "verify", "\\vf"))
         
        config.pluginsDirs.foreach { PluginManager.loadDir }
        
        // Run
        Console.initPassword
        Console.groovysh.run()
      } 
    } 
    IApplication.EXIT_OK
  }
  
  override def stop = {}
}
